docs - multibyte delimiter support for csv

lisakowen opened the pull request:
docs for https://github.com/greenplum-db/pxf/pull/956.

the `*:text` and `*:csv profiles` support now multi-byte and multiple character delimiters.  in this PR:
- add a topic "About Reading Data Containing Multi-Byte or Multi-Character Delimiters" to the hdfs page to describe the new feature.
- add a few example CREATE EXTERNAL TABLE calls, didn't create a step-by-step

notes:
- CREATE EXTERNAL TABLE supports DELIMITER/QUOTE/ESCAPE/NEWLINE.  this new formatter does as well, but it is based on the COPY framework.  COPY supports these as well.  i tried to call out what was different with this new formatter, not sure if i covered everything.
- it didn't seem to me that we needed any upgrade doc modifications, we already instruct the user to `ALTER EXTENSION pxf UPDATE` after they `pxf cluster register`.
- when the content is approved, i will add a parallel section or xref in the object store docs.

doc review site link (behind vpn):  https://docs-staging.vmware.com/en/VMware-Tanzu-Greenplum-Platform-Extension-Framework/mbdelim/tanzu-greenplum-platform-extension-framework/GUID-hdfs_text.html#about-reading-data-containing-multibyte-or-multicharacter-delimiters-7

denalex commented on docs/content/hdfs_text.html.md.erb:
````diff
@@ -365,3 +365,97 @@ Perform the following procedure to create Greenplum Database writable external t
+CREATE READABLE EXTERNAL TABLE byterep_delim (id int, city text, state text)
+  LOCATION ('pxf://multibyte_example?PROFILE=hdfs:text')
+FORMAT 'CUSTOM' (FORMATTER='pxfdelimited_import', DELIMITER=E'\xC2\xA4') ENCODING 'LATIN1';
+```
````
it is not really clear when this needs to be done. I think it looks like a user needs to do `E'..'` style when the table encoding is different from the database encoding ? If this is the case we need to clearly state this.

ashuka24 replied:
It is not required, a user can still use regular characters if there is a diff table encoding. However, if they choose to do byte representation, they must do it in the server encoding.

denalex commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,97 @@ Perform the following procedure to create Greenplum Database writable external t
+
+You can directly specify the delimiter or provide its byte representation.
+
+When you directly specify the delimiter, Greenplum Database converts between the file encoding and the database encoding for you. The following example directly specifies the `¤` currency symbol for the delimiter:
```
`When you directly specify the delimiter, Greenplum Database converts between the file encoding and the database encoding for you.` -- I am not sure what implications of this are. Is it good or bad ?

ashuka24 replied:
If you write `¤` or if you write `停` in the table definition, GPDB saves it internally in database encoding. 

Maybe it is better to say that Greenplum Database converts the option provided in the table definition into the database encoding

denalex commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,97 @@ Perform the following procedure to create Greenplum Database writable external t
+
+If you do not specify a quotation character for CSV data, PXF assumes that all columns are unquoted.
+
+When you specify a quotation character, PXF expects that all columns are quoted, and you must also specify an escape character. PXF reads any content between quote characters as-is, except for escaped characters.
```
`you must also specify an escape character.` - this is not what the example below shows, where no escape character is explicitly specified.

denalex commented on docs/content/hdfs_text.html.md.erb:
````diff
@@ -365,3 +365,97 @@ Perform the following procedure to create Greenplum Database writable external t
+"hello, my name is jane" she said. let's escape something \
+```
+
+When you specify an escape character and do not specify a quote character, PXF assumes that all columns are unquoted and escapes only the delimiter, newline, and escape itself.
````
in the example above `specify an escape character and do not specify a quote character` was true but all the columns were quoted, whereas the rest of the sentence says `PXF assumes that all columns are unquoted`

denalex commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,97 @@ Perform the following procedure to create Greenplum Database writable external t
+| Option Name  | Description | Default Value |
+|-------|---------------|---------------|
+| DELIMITER | The single-byte or multi-byte delimiter string that separates columns. The string may be up to 32 bytes in length, and may not contain quote or escape characters. **Required** | None |
+| QUOTE | When reading CSV data with a `*:csv` profile, the single one-byte ASCII quotation character for all columns. | `"` (double quote) |
```
what about `*:text` profile ?

also, what about null values ? For CSV it uses no value between delimiters, while for TEXT it uses `\N`. How does this formatter handle it ? Is there a way to specify a null value ?

lisakowen replied:
decided that this feature is for csv profiles only.

lisakowen commented:
hold off on reviewing, i am going to revise the docs a bit.

lisakowen commented:
@greenplum-db/gpdb-ud - ok, i pushed some changes to this topic and re-pushed the review site, please refresh.  i re-orged a bit, reframed some of the original content, and added an example section.  also added a level 2 topic about setting the table encoding (not specific to multibyte delimiter).

since setting the byte representation of the delimiter isn't required, i added it as a sub-section of the "setting the delimiter" section.  wondering if it is unnecessarily complicating matters, should i just remove it?

ashuka24 commented on docs/content/hdfs_text.html.md.erb:
````diff
@@ -365,3 +365,134 @@ Perform the following procedure to create Greenplum Database writable external t
+For example, if the database encoding is `UTF8`, the file encoding is `LATIN1`, and the delimiter is the `¤` currency symbol:
+
+```
+CREATE READABLE EXTERNAL TABLE byterep_delim (id int, city text, state text)
````
not sure if it was intended to have 2 create statements here, but the first table name is `byterep_delim` but the delimiter is not in the byte representation.

lisakowen replied:
nope, i didn't want to create statements.  i kept the second one.

ashuka24 commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,134 @@ Perform the following procedure to create Greenplum Database writable external t
 
+## <a id="about_encoding"></a>About Setting the External Table Encoding
+
+When the external file encoding differs from the database encoding, you must set the external table `ENCODING` to match that of the data file, create the external table as follows.
```
```suggestion
When the external file encoding differs from the database encoding, you must set the external table `ENCODING` to match that of the data file, create the external table as follows:

ashuka24 commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,134 @@ Perform the following procedure to create Greenplum Database writable external t
+| DELIMITER=\<delim_string\> | The single-byte or multi-byte delimiter string that separates columns. The string may be up to 32 bytes in length, and may not contain quote or escape characters. **Required** | None |
+| QUOTE=\<char\> | The single one-byte ASCII quotation character for all columns. | None |
+| ESCAPE=\<char\> | The single one-byte ASCII character used to escape special characters (for example, the `DELIM`, `QUOTE`,  or `NEWLINE` value, or the `ESCAPE` value itself). | None, or the `QUOTE` value if that is set |
+| NEWLINE=\<bytecode\> | The end-of-line indicator that designates the end of a row. Every line in the file must be terminated with the specified newline value. Valid values are `LF` (line feed, `0x0A`), `CRLF` (carriage return, `0x0D`), or `CR` (carriage return plus line feed, `0x0D` `0x0A`). | `LF` |
```
can we add a note here that if they set newline in the format options, it must also be set in the location uri?

lisakowen replied:
i didn't realize this.  is the syntax/behaviour the same as for the *:fixedwidth profile?  i.e. for that  profile, the docs state that the user sets the URI NEWLINE to CR/LF/CRLF and the formatter option (which is named line_delim for fixedwith and NEWLINE for this custom formatter) to something like `E'\n'`.

ashuka24 replied:
The behavior is the same. Instead of matching to what the fixedwidth profile has (`line_delim`) we intentionally decided to match with the existing CSV option `NEWLINE` which uses LF, CR, CRLF.

ashuka24 commented on docs/content/hdfs_text.html.md.erb:
````diff
@@ -365,3 +365,134 @@ Perform the following procedure to create Greenplum Database writable external t
+
+Given this external table definition, PXF reads the `sentence` text field as:
+
+```
````
this may be more clear if you did a `SELECT sentence FROM mbyte_delim_quoted_escaped`

ashuka24 commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,134 @@ Perform the following procedure to create Greenplum Database writable external t
+CREATE READABLE EXTERNAL TABLE byterep_delim (id int, city text, state text)
+  LOCATION ('pxf://multibyte_example?PROFILE=hdfs:csv')
+FORMAT 'CUSTOM' (FORMATTER='pxfdelimited_import', DELIMITER='¤') ENCODING 'LATIN1';
+CREATE EXTERNAL TABLE pxf_csv_latin1(location text, month text, num_orders int, total_sales float8)
+  LOCATION ('pxf://data/pxf_examples/pxf_hdfs_simple.txt?PROFILE=hdfs:csv')
+FORMAT 'CSV' ENCODING 'LATIN1';
```
```suggestion
```

i think if we want to keep this, we should not have the line on 372

yimingli-vmware commented on docs/content/hdfs_text.html.md.erb:
```diff
@@ -365,3 +365,149 @@ Perform the following procedure to create Greenplum Database writable external t
+| Yes<sup>2</sup> | Yes (the default is the quote value). | PXF reads data between the quotes as-is except for escaped characters. |
+| Yes<sup>2</sup> | No (`ESCAPE 'OFF'`) | PXF reads the data between quote characters as-is. |
+| No<sup>1</sup> | Yes | PXF reads the data as-is and un-escapes only the delimiter, newline, and escape itself. |
+| Yes<sup>2</sup> | Yes | PXF reads the data between quote characters as-is and un-escapes only the quote and escape characters. |
```
```suggestion
| Yes<sup>2</sup> | Yes | PXF reads the data between quote characters as-is and un-escapes only the quote and escaped characters. |
```

What's the difference between line 445 and line 442?

ashuka24 replied:
for line 442, i think it is the situation where escape is not defined in the table definition. but we set a default value since quote is defined. i think this could be clarified a bit more to mitigate this confusion @lisakowen 

here in line 445, we are referring to the character set as quote and the character set as escape. so `the quote and escape characters` is actually the correct phrase.

lisakowen commented:
thank you for reviewing @ashuka24 and @yimingli-vmware .
i am merging this PR.  will separately address any further comments.
